### PR TITLE
Resolved bug (8554) with Task Based Ping API

### PR DESCRIPTION
**Pull request by amccague**
Resolved bug (8554) with Task Based Ping API  …

`System.Net.NetworkInformation.Ping`

https://bugzilla.xamarin.com/show_bug.cgi?id=8554

Added tests to cover the Asynchronous API and the Task Based API.

Fixed an issue with the Task Based API:

- `public Task<PingReply> SendPingAsync (IPAddress address)`
- `public Task<PingReply> SendPingAsync (string hostNameOrAddress)`
- etc

that would previously throw a `NullReferenceException` when invoked as the class' `CancellationTokenSource` was not instantiated when getting `cts.Token`.

The `CancellationTokenSource` was also not disposed of after use.

### CODE REVIEW

**Comment by akoeplinger** [comment]
@monojenkins build

**Comment by akoeplinger** [comment]
build

File: mcs/class/System/System.Net.NetworkInformation/Ping.cs [hunk 1]
```diff
@@ -159,6 +159,8 @@ protected void OnPingCompleted (PingCompletedEventArgs e)
 		{
 			user_async_state = null;
 			worker = null;
+
+			cts?.Dispose();
```
Comment (marek-safar) [inline mcs/class/System/System.Net.NetworkInformation/Ping.cs]: You cannot use ?. here (we allow bootstrap with old versions of mcs)

Comment (amccague) [inline mcs/class/System/System.Net.NetworkInformation/Ping.cs]: Sure, same for tests I presume?

Comment (marek-safar) [inline mcs/class/System/System.Net.NetworkInformation/Ping.cs]: No, tests are ok

**Pull request merged**